feat: enable custom sections

sircharlo opened the pull request:
To do:

- [x] Create dialog to show existing custom sections and allow adding new ones
- [x] Set random color when creating new section
- [x] Add [color picker](https://quasar.dev/vue-components/color-picker/)
- [x] Modify logic to include custom sections in the media list
- [x] Allow sorting of custom sections
- [x] Allow drag-dropping to and from custom sections
- [x] Properly persist drag/drop reordering and list changing for custom sections
- [x] Show media items in standard "additional" section if the referenced section doesn't exist anymore
- [x] Add new section button to bottom of media list
- [x] Allow renaming the standard "additional" section, changing its color and reordering it 
- [x] Rename 'Extra media' into something more generic
- [x] Fix the "New section" button fill and hover color in dark mode
- [x] Adding media to a specific section (by using the + icon) still adds the media to "Extra media" section and requires manual drag-n-drop sorting; fix this.
- [x] Enable section renaming by double clicking the name
- [x] Optimize section renaming and color changing
- [x] Move "reorder sections" button to the top menu bar
- [x] Add "move up" and "move down" buttons to sections to quickly move them 
- [x] Update drag handle icon with the one in Figma
- [x] Fix the reorder sections logic
- [x] Fix "reset media order"
- [x] Hide extra action (all except for 'add media' to "more" menu for clean UI
- [x] Look into adding subheadings/dividers
- [x] Add 'repeat section' option
- [x] Remove debugging text
- [x] Fix sorting

#### Color picker
- [x] RGB and HEX top tabs are not really necessary; remove complicated header
- [x] Remove header and footer completely
- [x] Change color icon with the one in Figma

Unsure:
- [ ] Allow sorting sections in main media list
- [ ] Remove dialog if unneeded
- [ ] Enable heading/section import from jwpub file

Related #3479

sircharlo commented on src/pages/MediaCalendarPage.vue:
```diff
@@ -68,22 +68,31 @@
       />
+      <template
+        v-for="mediaList in mediaLists"
+        :key="mediaList.items.map((m) => m.uniqueId).join(',')"
```
Maybe this is the reason of the flicker on sort?

sircharlo commented:
@ux-git Any thoughts about this build?

I think for the sake simplicity and consistency, it might be best to fold the color/delete/reorder buttons that are currently in the section headers into an 'edit' or 'more' button. This button would open the dialog where users could do all those things.

I'm not decided yet though. I see this going a few potential ways. I want a mix of customization possibilities with simplicity for non-power users

ux-git commented:
@sircharlo very good so far!

Some thoughts: 
0. All the media for the special events are 'Extra media' basically, since there are no standard media. What about renaming the 'Extra media' into something more generic like 'Added media' or 'Imported media'?
1. **Color picker:** 
a) RGB and HEX top tabs are not really necessary I believe. Since we have the bottom tabs with different options, moreover RGB vs HEX currently changes nothing besides color code input. I'm not even sure if users will benefit from seeing the color codes.
b) RGB sliders can be simply removed – it's one of the most unpredictable way of selecting a color 😄😅 HEX palette and presets grid covers literally all the cases that I can imagine.
c) This area in the middle of the palette is an eye-burning-sector. I personally would remove it if possible for the sake of not being able to make the app look bad. Other than those, there are more than 50 other colors which is more than enough.
<img width="178" alt="Screenshot 2025-01-23 at 20 09 33" src="https://github.com/user-attachments/assets/f7e7b840-6df8-4381-87fe-9581bc168726" /><br>
d) The color icon would be nice to change to something like in Figma, e.g. palette, color droplet, color picker, or something else. Right now the dot is colored in the same way as other icons do, so it doesn't communicate its purpose right away.
2. New section button (in the reorder 'Edit sections' modal as well) needs an alternate fill and hover color for the dark theme: 
<img width="873" alt="Screenshot 2025-01-23 at 20 14 26" src="https://github.com/user-attachments/assets/26f4486b-209a-4461-aa3c-de70d8545aa6" /><br>
3. Adding media to a specific section (by using the + icon) still adds the media to "Extra media" section and requires manual drag-n-drop sorting.
4. Would it be possible to rename sections by double clicking on the name? It would be a great addition to the global sorting modal window.
5. Renaming the sections is quite slow, perhaps because of the real-time update.
6. **Section reordering:**
a) Currently, every section has the same button which opens the modal window with global reorder settings. Maybe this could be replaced with the 'Reorder' button in the blue top bar? One button for all sections make more sense than many buttons for each section that are doing the same thing.
b) If we remove the sorting shortcut, we can actually place 'up' and 'down' buttons there to quickly move the section one step higher or lower with just one click. For reordering something for huge set of sections global reorder window is the way to go, but for quick small edits this might save a couple of clicks for users.
c) The drag handle can be a bit more intuitive for non-power users by showing up/down arrows on it to let visually know that it can be dragged and that's not an additional menu. You can find this and some other icons in Figma. If others are needed, don't hesitate to request some:
<img width="86" alt="Screenshot 2025-01-23 at 21 26 41" src="https://github.com/user-attachments/assets/a4ffcf1d-0e64-4989-825f-0e5fbacf697d" /><br>

7. Regarding the 'More' button... I see many users requesting 'one-click-away' solutions, and I can see the existing set of buttons as a nice additional color indicator for the right side of each section. However, hiding all the buttons inside the ... menu is an attractive idea for me personally. Perhaps I would leave only the ADD MEDIA button to make it easily accessible.
8. I don't know if it's a macOS only thing, or just a temporary state of code, but there are code snippets all over the place (on the normal meetings profile as well), just want to let you know:
<img width="882" alt="Screenshot 2025-01-23 at 21 26 25" src="https://github.com/user-attachments/assets/e152fa5d-c9aa-41dd-bed2-4ec6ac834ec0" />
9. Why are you unsure about the 'Enable heading/section import from jwpub file'? 
10. What do you think about an option of adding subheadings into each section? Those would serve as dividers and can drastically multiply the number of covered use cases. For instance, I can see a section as a 'Pioneer school: Part 7B' and subheadings inside can indicate the topics discussed with related media. Alternatively, for a theoretical assembly use case it's 'Morning session' and 'Afternoon session' sections with public talks listed as subheadings. It's just an add-on idea to make the whole system more flexible and with more precise fine-tuning possibilities. Many jwpub files have similar structure: [Sub]heading + media list, then divider and next [sub]heading + media list within one file (e.g. Pioneer schooling playlist, CO visit playlist, I'm sure I've seen some others too).
But of course, you're the one to decide whether it is worth trying, or better to start with MVP and then keep this idea for later in case if users really request something like this.

e-Feitosa commented:
Did not find in beta this implementation, or has not yet been fully finalized to be beta?

ux-git commented:
@e-Feitosa [Build #5733](https://github.com/sircharlo/meeting-media-manager/actions/runs/13175022861)

e-Feitosa commented:
> @e-Feitosa [Build #5733](https://github.com/sircharlo/meeting-media-manager/actions/runs/13175022861)

It's looking excellent, great job! One question:
Will it be possible to drag media from one section to another? Or is the idea that if media is added to a section, for example: MORNING and AFTERNOON sections, I can only move the media within the MORNING section? Or will it be possible to move media from the MORNING section to the AFTERNOON section?

mtdvlpr commented:
> > @e-Feitosa [Build #5733](https://github.com/sircharlo/meeting-media-manager/actions/runs/13175022861)
> 
> It's looking excellent, great job! One question: Will it be possible to drag media from one section to another? Or is the idea that if media is added to a section, for example: MORNING and AFTERNOON sections, I can only move the media within the MORNING section? Or will it be possible to move media from the MORNING section to the AFTERNOON section?

I believe it should behave similarly to the built-in LAM and WT sections, so you should be able to move media between sections as well as within sections.

sircharlo commented:
> @sircharlo very good so far!

Thank you! I finally think I'll have a bit of time to look into this PR over the coming weeks.
 
> Some thoughts: 0. All the media for the special events are 'Extra media' basically, since there are no standard media. What about renaming the 'Extra media' into something more generic like 'Added media' or 'Imported media'?

Done.

> 1. **Color picker:**
>    a) RGB and HEX top tabs are not really necessary I believe. Since we have the bottom tabs with different options, moreover RGB vs HEX currently changes nothing besides color code input. I'm not even sure if users will benefit from seeing the color codes.

Done. 

>    b) RGB sliders can be simply removed – it's one of the most unpredictable way of selecting a color 😄😅 HEX palette and presets grid covers literally all the cases that I can imagine.

Done.

>    c) This area in the middle of the palette is an eye-burning-sector. I personally would remove it if possible for the sake of not being able to make the app look bad. Other than those, there are more than 50 other colors which is more than enough.

We can specify a pallette, actually ([example](https://quasar.dev/vue-components/color-picker#custom-palette)). Any suggestions?

>    d) The color icon would be nice to change to something like in Figma, e.g. palette, color droplet, color picker, or something else. Right now the dot is colored in the same way as other icons do, so it doesn't communicate its purpose right away.

Agreed. Feel free to make an icon that would be clear.

> 2. New section button (in the reorder 'Edit sections' modal as well) needs an alternate fill and hover color for the dark theme:

I plan to fix this sometime soon.

> 3. Adding media to a specific section (by using the + icon) still adds the media to "Extra media" section and requires manual drag-n-drop sorting.

I plan to fix this sometime soon.

> 4. Would it be possible to rename sections by double clicking on the name? It would be a great addition to the global sorting modal window.

I plan to add this sometime soon.

> 5. Renaming the sections is quite slow, perhaps because of the real-time update.

I plan to fix this sometime soon.

> 6. **Section reordering:**
>    a) Currently, every section has the same button which opens the modal window with global reorder settings. Maybe this could be replaced with the 'Reorder' button in the blue top bar? One button for all sections make more sense than many buttons for each section that are doing the same thing.

Good call; I plan to implement this sometime soon.

>    b) If we remove the sorting shortcut, we can actually place separate 'up' and 'down' buttons there to quickly move the section one step higher or lower with just one click. For reordering something for huge set of sections global reorder window is the way to go, but for quick small edits this might save a couple of clicks for users.

Nice idea.

>    c) The drag handle can be a bit more intuitive for non-power users by showing up/down arrows on it to let visually know that it can be dragged and that's not an additional menu. You can find this and some other icons in Figma. 

I plan to fix this sometime soon.

> 7. Regarding the 'More' button... I see many users requesting 'one-click-away' solutions, and I can see the existing set of buttons as a nice additional color indicator for the right side of each section. However, hiding all the buttons inside the ... menu is an attractive idea for me personally. Perhaps I would leave only the ADD MEDIA button to make it easily accessible.

I like this idea.

> 8. I don't know if it's a macOS only thing, or just a temporary state of code, but there are code snippets all over the place (on the normal meetings profile as well), just want to let you know

Yeah, this is just for debugging the new custom section logic. It won't make it into the production version.

> 9. Why are you unsure about the 'Enable heading/section import from jwpub file'?

From what I've seen, it's not cuurently technically feasible to extract that kind of data reliably from JWPUB files, unfortunately.

> 10. What do you think about an option of adding subheadings into each section? Those would serve as dividers and can drastically multiply the number of covered use cases. For instance, I can see a section as a 'Pioneer school: Part 7B' and subheadings inside can indicate the topics discussed with related media. Alternatively, for a theoretical assembly use case it's 'Morning session' and 'Afternoon session' sections with public talks listed as subheadings. It's just an add-on idea to make the whole system more flexible and with more precise fine-tuning possibilities. Many jwpub files have similar structure: [Sub]heading + media list, then divider and next [sub]heading + media list within one file (e.g. Pioneer schooling playlist, CO visit playlist, I'm sure I've seen some others too).

I'll think about this one.

> 12. Also, what about Repeat feature for the whole section? For cases where multiple videos must be played in a loop. Not even necessarily for a convention/assembly which come first to my mind, from my experience this is often done for funeral talk/event to display videos after the program until all the people leave.

I'll think about this one, too.

ux-git commented:
> We can specify a pallette, actually ([example](https://quasar.dev/vue-components/color-picker#custom-palette)). Any suggestions?

This one looks simply enough for users: 
<img width="267" alt="Screenshot 2025-07-06 at 14 47 10" src="https://github.com/user-attachments/assets/5e652ed7-3dbc-4370-bee5-7327f4bf33b3" />

> Agreed. Feel free to make an icon that would be clear.
Already available in Figma:
<img width="54" alt="Screenshot 2025-07-06 at 14 49 56" src="https://github.com/user-attachments/assets/a6b7e4b1-770b-44ce-a1a1-4c6f23594eb2" />

sircharlo commented:
@ux-git I'd like your feedback again for this PR. I've made alot of progress, and I dare say it's almost ready for release.

ux-git commented:
@sircharlo great news! I'll take a look in a few days at best, as I have the convention. I'll drop my feedback here ⚡

sircharlo commented:
> @sircharlo great news! I'll take a look in a few days at best, as I have the convention. I'll drop my feedback here ⚡

Enjoy the convention!!

ux-git commented:
> > @sircharlo great news! I'll take a look in a few days at best, as I have the convention. I'll drop my feedback here ⚡
> 
> Enjoy the convention!!

Thanks!

ux-git commented:
@sircharlo would you be able to attach a Mac version as well after the next few days, once you have a build that is quite release-ready? I'm not sure I'll have a Windows machine to test it out during upcoming weeks

ux-git commented:
@sircharlo yes, it works for me

sircharlo commented:
> @sircharlo yes, it works for me

Perfect! Let me know your thoughts 😄

sircharlo commented:
> 4. I guess the top right corner of a divider supposed to be sharp to follow the design language?

Fixed!


> 5. The text field and buttons in this modal window are getting active only after a few seconds, at first they are not responding to clicks:

This should be better now, I've refactored a few things in the latest test build to improve performance. Let me know.

> 6. What's the intended behaviour of the blue dashed outline? What should it indicate?
>    Right now it just shifts the layout by its width size, but I'm not sure if it provides useful guidance for users as it's not drag and drop, and the same mechanics on the normal meetings page work with no borders like this. Am I missing something?

It's just a visual indicator of where to drop media while sorting. Do you think it's unneccesary?

> 7. If an image is included into the repeated section, it's being demonstrated for 10 seconds with no indicators or way to adjust that. Maybe a timeline for images would be meaningful there? Or a dropdown with a time selection on the left of the Play button (with 10s, 30s, 1m, 5m, 10m)?

Good idea, but I'd like to add that functionality a bit later. I want to release this as soon as possible, so could you maybe open an issue so that I could tackle that for the next release?

> 8. Would it be possible to keep the divider's icon the same size as the section's ones? Just to aim easier. You could compensate the top-bottom padding size by making it smaller to avoid having too big divider. Maybe I'd reduce even the current height by 8-12px just to make it more of a secondary element giving it less optical weight.

Same as point 7.

> 9. Overall, a really awesome work done, congrats! I'm just wondering how we could give this powerful feature a bit more attention? Right now to access it user has to select "Other" meeting type while creating a congregation profile. Do you think we could you something more prominent? For example give this mode its own name like 'Special events' instead of being 'other'. Other alone sounds like a simplified version of the main functionality, but in this case it's even more advanced. For sure, this feature stays secondary, but we could provide more context about what it can do t least. Do you have any better ideas?

Thank you! I've revamped the docs website completely; it's almost ready to go, but I'm waiting to release this. The docs will be much more complete now, and hopefully help users set up what they need faster and easier. If you can think of a better way to indicate that this is available, let me know. Eventually, I'd like to add a "did you know" popup on startup with practical tips for users, as I think this might be a good way to discover features.

ux-git commented:
> It's just a visual indicator of where to drop media while sorting. Do you think it's unneccesary?

I believe that since the media item can be inserted literally anywhere (the same as in the normal meeting mode), showing it won't help much. I think without the border it will be cleaner yet still clear how it works. What do you think?

> could you maybe open an issue so that I could tackle that for the next release?

Sure thing👌 

> Eventually, I'd like to add a "did you know" popup on startup with practical tips for users, as I think this might be a good way to discover features.

Sounds good! A quick intro to the app, briefly describing the core features with minimal illustrations would be a nice way to improve the onboarding.

sircharlo commented:
> > It's just a visual indicator of where to drop media while sorting. Do you think it's unneccesary?
> 
> I believe that since the media item can be inserted literally anywhere (the same as in the normal meeting mode), showing it won't help much. I think without the border it will be cleaner yet still clear how it works. What do you think?

It's more the fact that dropping it on a header itself won't reorder the item; it's more to indicate to the user where the dropzone is. Does that make sense?

> > could you maybe open an issue so that I could tackle that for the next release?
> 
> Sure thing👌

Thanks!

> > Eventually, I'd like to add a "did you know" popup on startup with practical tips for users, as I think this might be a good way to discover features.
> 
> Sounds good! A quick intro to the app, briefly describing the core features with minimal illustrations would be a nice way to improve the onboarding.

I agree!

ux-git commented:
> It's more the fact that dropping it on a header itself won't reorder the item; it's more to indicate to the user where the dropzone is. Does that make sense?

I got the point right away, but in our case the huge hint for user is the pre-drop animation that instantly shows where the media item will be as soon as you release the mouse button. So to me it's totally intuitive, and adding the dashed border is a duplicate. What do you think about starting with the non-border version, and adding it if you'll get any feedback of this being not clear enough?

ux-git commented:
@sircharlo deal
